CONTRIBUTING.md

```diff
@@ -19,17 +36,223 @@ Follow either of the two links above to access the appropriate CLA and
+**region tags**.
+
+Example:
+
+```ruby
+def create_bucket project_id:, bucket_name:
+  # [START create_bucket]
+  # project_id  = "Your Google Cloud project ID"
+  # bucket_name = "Your Google Cloud Storage bucket name"
+  
+  require "google/cloud/storage"
+  
+  storage = Google::Cloud::Storage.new project: project_id
+  
+  bucket = storage.create_bucket bucket_name
+  
+  puts "Created bucket #{bucket.name}"
+  # [END create_bucket]
+end
+```
+
+### Reference snippets
+
+Reference snippets are isolated snippets that demonstrate how to
+perform a specific task such as creating a Google Cloud Storage bucket.
+
+Ideally, developers should be able to **copy/paste** the code in a reference
+snippet and run it after replacing placeholder values.
+
+#### Placeholder values
+
+When a code snippet makes use of variables that are defined outside
+of the snippet, add commented-out variable declarations to the top
+of the snippet.
+
+Example:
+
+```ruby
+# project_id  = "Your Google Cloud project ID"
+# bucket_name = "Your Google Cloud Storage bucket name"
+  
+require "google/cloud/storage"
+  
+storage = Google::Cloud::Storage.new project: project_id
+```
+
+#### require statements
+
+Reference snippets should include the `require` statement(s) necessary
+to import relevant dependencies for the code sample.
+
+If a snippet uses a Google Cloud client library, it should be required
+in the snippet.  This documents how to require the client library.
+
+Example:
+
+```ruby
+# project_id  = "Your Google Cloud project ID"
+# bucket_name = "Your Google Cloud Storage bucket name"
+
+require "google/cloud/storage"
+
+storage = Google::Cloud::Storage.new project: project_id
+```
+
+**Note:** common libraries such as "date" and "time" do not need to be
+required within the snippet.  Explicitly require dependencies that
+add value and should be documentated.
+
+### Executing snippets
+
+When possible, reference code snippets should be executable.
+
+Each reference code snippet should be executable via a
+corresponding command-line application.
+
+For example, the "create_bucket" snippet in "buckets.rb" should be
+executable by running `bundle exec ruby buckets.rb create`
+
+#### Argument parsing
+
+To make snippets executable, add a postable to the end of the file
+that calls the snippet's method.
+
+For example:
+
+```ruby
+def create_bucket project_id:, bucket_name:
+  # ...
+end
+
+if __FILE__ == $PROGRAM_NAME
+  command = ARGV.shift
+  
+  case command
+  when "create"
+    create_bucket project_id:  ENV["GOOGLE_CLOUD_PROJECT"],
+                  bucket_name: ARGV.first
+  else
+   puts <<-usage
+Usage: bundle exec ruby buckets.rb [command] [arguments]
+
+Commands:
+  create <bucket>    Create a new bucket with the provided name
+  
+Environment variables:
+  GOOGLE_CLOUD_PROJECT must be set to your Google Cloud project ID
+   usage
+  end
+end
+```
+
+Argument parsing is intentionally simple.
+
+### Tutorial applications
+
```

Comment:
Should we mention the path **appengine/** holds Tutorial applications? It could provide more examples.